File: src/algorithm/String/FractionToDecimal166.java
package algorithm.String;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * 166. 分数到小数
 * 给定两个整数，分别表示分数的分子 numerator 和分母 denominator，以 字符串形式返回小数 。
 * 如果小数部分为循环小数，则将循环的部分括在括号内。
 * 如果存在多个答案，只需返回 任意一个 。
 * 对于所有给定的输入，保证 答案字符串的长度小于 104 。
 *
 * 示例 1：
 * 输入：numerator = 1, denominator = 2
 * 输出："0.5"
 * 示例 2：
 * 输入：numerator = 2, denominator = 1
 * 输出："2"
 * 示例 3：
 * 输入：numerator = 4, denominator = 333
 * 输出："0.(012)"
 *
 * 提示：
 * -231 <= numerator, denominator <= 231 - 1
 * denominator != 0
 * */
public class FractionToDecimal166 {
    public static void main(String[] args) {
        int min=Integer.MIN_VALUE;
        int mod=-1;
        //溢出
        long res=min/mod;
        long min1=min,mod1=mod;
        //没有溢出
        long res1=min1/mod1;
        System.out.printf("res::%s,res1:%s",res,res1);
    }

    public String fractionToDecimal(int numerator, int denominator) {
        // 转 long 计算，防止溢出 如果是两个int相除，即使用long接收结果仍然也可能溢出
        long a = numerator, b = denominator;
        // 如果本身能够整除，直接返回计算结果
        if (a % b == 0) return String.valueOf(a / b);
        StringBuilder sb = new StringBuilder();
        // 如果其一为负数，先追加负号
        if (a * b < 0) sb.append('-');
        a = Math.abs(a); b = Math.abs(b);
        // 计算小数点前的部分，并将余数赋值给 a
        sb.append(a / b + ".");
        a %= b;
        Map<Long, Integer> map = new HashMap<>();
        while (a != 0) {
            // 记录当前余数所在答案的位置，并继续模拟除法运算
            map.put(a, sb.length());
            a *= 10;
            sb.append(a / b);
            a %= b;
            // 如果当前余数之前出现过，则将 [出现位置 到 当前位置] 的部分抠出来（循环小数部分）
            if (map.containsKey(a)) {
                int u = map.get(a);
                return String.format("%s(%s)", sb.substring(0, u), sb.substring(u));
            }
        }
        return sb.toString();
    }

}

File: go/src/String/FractionToDecimal166.go
package main

import (
	"fmt"
	"strconv"
)

// 由于Golang中的int的取值是和和操作系统位数相关的，如果是32位操作系统，int类型的大小就是4字节;
// 如果是64位操作系统，int类型的大小就是8个字节。所以我们在这里就不用担心int相除数据溢出的问题了

// 这里的拼接字符串采用的是byte数组  太麻烦了
func fractionToDecimal(numerator int, denominator int) string {
	if numerator%denominator == 0 {
		return strconv.Itoa(numerator / denominator)
	}
	res := []byte{}
	if numerator*denominator < 0 {
		res = append(res, '-')
	}
	a, b := abs(numerator), abs(denominator)
	m1 := map[int]int{}
	res = append(res, []byte(strconv.Itoa(a/b))...)
	res = append(res, '.')
	less := a % b
	for less != 0 {
		m1[less] = len(res)
		less *= 10
		res = append(res, []byte(strconv.Itoa(less/b))...)
		less %= b
		if val, ok := m1[less]; ok {
			return fmt.Sprintf("%s(%s)", string(res[:val]), string(res[val:]))
		}
	}
	return string(res)
}

// 这里的拼接字符串采用的是 + 号
func fractionToDecimal2(numerator int, denominator int) string {
	if numerator%denominator == 0 {
		return strconv.Itoa(numerator / denominator)
	}
	res := ""
	if numerator*denominator < 0 {
		res += "-"
	}
	a, b := abs(numerator), abs(denominator)
	m1 := map[int]int{}
	res += strconv.Itoa(a/b) + "."
	less := a % b
	for less != 0 {
		m1[less] = len(res)
		less *= 10
		res += strconv.Itoa(less / b)
		less %= b
		if val, ok := m1[less]; ok {
			return fmt.Sprintf("%s(%s)", res[:val], res[val:])
		}
	}
	return res
}

func abs(val int) int {
	if val < 0 {
		return -val
	}
	return val
}
